crossdock/client/gauntlet/behavior.go

```diff
@@ -0,0 +1,485 @@
+		case "", "ThriftTest":
+			client = reflect.ValueOf(thrifttestclient.New(channel))
+		case "SecondService":
+			client = reflect.ValueOf(secondserviceclient.New(channel))
+		default:
+			assert.Fail("", "%v: unknown thrift service", desc)
+		}
+
+		f := client.MethodByName(tt.function)
+		if !checks.True(f.IsValid(), "%v: invalid function", desc) {
+			continue
+		}
+		ft := f.Type()
+
+		if !checks.Equal(1+len(tt.give), ft.NumIn(), "%v: should accept %d arguments", desc, 1+len(tt.give)) {
+			continue
+		}
+
+		if tt.want != nil {
+			if !checks.Equal(3, ft.NumOut(), "%v: should return 3 values", desc) {
+				continue
+			}
+		}
+
+		ctx, _ := context.WithTimeout(context.Background(), time.Second)
+		req := thrift.Request{
+			Context: ctx,
+			TTL:     time.Second, // TODO context TTL should be enough
+		}
+
+		args := []reflect.Value{reflect.ValueOf(&req)}
+		for i, v := range tt.give {
+			var val reflect.Value
+			vt := ft.In(i + 1)
+			if v == nil {
+				// nil is an invalid argument to ValueOf. For nil, use the zero
+				// value for that argument.
+				val = reflect.Zero(vt)
+			} else {
+				val = reflect.ValueOf(v)
+			}
+			if !checks.Equal(vt, val.Type(), "%v: argument %v type mismatch", desc, i) {
+				continue
+			}
+			args = append(args, val)
+		}
+
+		var (
+			returns = f.Call(args)
+			err     error
+			got     interface{}
+		)
+
+		switch len(returns) {
+		case 2:
+			e := returns[1].Interface()
+			if e != nil {
+				err = e.(error)
+			}
+		case 3:
+			got = returns[0].Interface()
+			e := returns[2].Interface()
+			if e != nil {
+				err = e.(error)
+			}
+		default:
+			assert.Fail("",
+				"%v: received unexpected number of return values: %v", desc, returns)
+		}
+
+		if _, isBadRequest := err.(transport.BadRequestError); isBadRequest {
+			// TODO: Once all other languages implement the gauntlet test
+			// subject, we can remove this check.
+			if strings.Contains(err.Error(), "unrecognized procedure") {
+				behavior.Skipf(s, "%v: procedure not defined", desc)
+				continue
+			}
+		}
+
+		if tt.wantError != nil || tt.wantErrorLike != "" {
+			if !checks.Error(err, "%v: expected failure but got: %v", desc, got) {
+				continue
+			}
+			if tt.wantError != nil {
+				assert.Equal(tt.wantError, err, "%v: server returned error: %v", desc, err)
+			}
+			if tt.wantErrorLike != "" {
+				assert.Contains(err.Error(), tt.wantErrorLike, "%v: server returned error: %v", desc, err)
+			}
+		} else {
+			if !checks.NoError(err, "%v: call failed", desc) {
+				continue
+			}
+
+			if tt.want != nil {
+				assert.Equal(tt.want, got, "%v: server returned: %v", desc, got)
+			}
+		}
+	}
+}
+
+func boolp(x bool) *bool         { return &x }
+func bytep(x int8) *int8         { return &x }
+func int16p(x int16) *int16      { return &x }
+func int32p(x int32) *int32      { return &x }
+func int64p(x int64) *int64      { return &x }
+func doublep(x float64) *float64 { return &x }
+func stringp(x string) *string   { return &x }
```

Comment:
I don't really want a trivial "helpers" package. The thing they do is simple enough that we can create them on demand or just make use of local variables.